doc/source/OEP/OEP_002.rst

```diff
@@ -0,0 +1,239 @@
+                         get('streaming', None) == 'ondemand')]
+       want_to_stream = [d for d in devices
+                         if (getattr(d, 'read_mode', {}).
+                             get('streaming', None) == 'always')]
+
+       def inner_trigger_and_read():
+           grp = _short_uid('trigger')
+           no_wait = True
+           for obj in devices:
+               if hasattr(obj, 'trigger'):
+                   no_wait = False
+                   yield from trigger(obj, group=grp)
+           if want_to_stream:
+               yield from create(f'{name}_streaming')
+               for obj in want_to_stream + may_stream:
+                   yield from read(obj, read_mode='streaming')
+               yield from save()
+           # Skip 'wait' if none of the devices implemented a trigger method.
+           if not no_wait:
+               yield from wait(group=grp)
+           yield from create(name)
+           ret = {}  # collect and return readings to give plan access to them
+           for obj in devices:
+               reading = (yield from read(obj))
+               if reading is not None:
+                   ret.update(reading)
+           yield from save()
+           return ret
+       from .preprocessors import rewindable_wrapper
+       return (yield from rewindable_wrapper(inner_trigger_and_read(),
+
+
+Push dark-frame logic down into devices
+---------------------------------------
+
+This is informed by discussions with CJ
+
+To support analysis with dark frame it may be easier to always ensure
+that there is a 'dark' stream along side the 'primary' stream rather
+than taking single run dark frame (to keep analysis data-broker agnostic and
+to enable taking dark frames during long scans).  One way to achieve this is
+via an smarter inner step plan ::
+
+   def smart_one_1d_step(detectors, motor, step):
+
+       need_dark_frame = [d for d in detectors
+                          if ('dark' in getattr(d, 'read_modes', {}))]
+
+       def take_dark():
+           grp = _short_uid('trigger')
+           no_wait = True
+           for obj in need_dark_frame:
+               # there are skeletons buried here!
+               if is_dark_stale(obj):
+                   no_wait = False
+                   yield from trigger(obj, read_mode='dark', group=grp)
+
+           if no_wait:
+               return
+
+           yield from wait(group=grp)
+           yield from create('dark')
+           for obj in need_dark_frame:
+               yield from read(obj, read_mode='dark')
+           yield
+
+       def move():
+           grp = _short_uid('set')
+           yield Msg('checkpoint')
+           yield Msg('set', motor, step, group=grp)
+           yield from take_dark()
+           yield Msg('wait', None, group=grp)
+
+       yield from move()
+       return (yield from trigger_and_read(list(detectors) + [motor]))
+
+
+The ``is_dark_stale`` function is doing alot of work, we may need to
```

Comment:
Great example. As you suggested before in a previous conversation, read_mode can also be a function.
```
get_dark = det.getattr('read_mode', {}).get('dark', lambda : False)
if get_dark():
    # take dark
```
where `get_dark` could contain some of the logic.
This would at least move the function from some global function to an object member, which is safer. What do you think?

Reply:
I don't remember saying that, but it is a good idea!

Slightly concerned about the wildly varying semantics on the keys....

Reply:
you mentioned it when trying to emphasize the power of having a dictionary of keys pointing to arbitrary object (as opposed to `bool`s), that they could also be functions :-D